Add scene.delete service for dynamically created scenes (with scene.create)

tetele opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a new `scene.delete` service that deletes temporary scenes created with the `scene.create` service.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/26504

Although there is no open issue for this PR, here are a few links to it being requested:
- https://community.home-assistant.io/t/scene-delete-or-remove-service-call/425456
- https://community.home-assistant.io/t/delete-dynamic-scene/246208
- https://community.home-assistant.io/t/dynamic-scenes-ability-to-peek-inside-and-delete-scene/500203

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

tetele committed:
Added scene.delete service

Only for scenes created with scene.create

tetele commented:
Not stale. Will merge base branch when I'll be back from my holiday :D

sidevesh commented:
@tetele what is left for this to be merged ? Is this ready for review ? This has been stuck for quite some time.

tetele commented:
I see some failing tests after a rebase. I will look into it the next few days, but nobody has picked this up for review yet.

tetele committed:
Refactor after #95984 #96390

tetele commented:
There, @sidevesh, all tests are green. Now maybe someone can review the PR. Feel free to do so yourself, if you want to.

sidevesh commented:
thanks @tetele, I don't have write access so I can't review, @MartinHjelmare can you review maybe ?

MartinHjelmare commented:
Please don't ping for review.

rokam commented on homeassistant/components/homeassistant/scene.py:
```diff
@@ -271,6 +282,25 @@ async def create_service(call: ServiceCall) -> None:
+        if scene is None:
+            _LOGGER.error("The scene %s does not exist", entity_id)
+            return
+        if not isinstance(scene, HomeAssistantScene) or not scene.from_service:
```
I think this check should be splitted. One to check if the entity is a scene with its own log message and the other one with the actual msg.

rokam commented on homeassistant/components/homeassistant/scene.py:
```diff
@@ -271,6 +282,25 @@ async def create_service(call: ServiceCall) -> None:
 
+    async def delete_service(call: ServiceCall) -> None:
+        """Delete a dynamically created scene."""
+        entity_id = call.data[CONF_ENTITY_ID]
```
Why not use scene_id?

MartinHjelmare replied:
We should not use `scene_id`. The parameter `scene_id` in the create service isn't an entity_id. It's the name of the service to be created. When configuring scenes in YAML, this parameter is actually called `name`. I don't think we would have named this parameter `scene_id` today in the create service.

`entity_id` is good as we want an entity_id to find the existing scene entity.

rokam commented on homeassistant/components/homeassistant/scene.py:
```diff
@@ -84,6 +84,7 @@ def _ensure_no_intersection(value: dict[str, Any]) -> dict[str, Any]:
 
 
 CONF_SCENE_ID = "scene_id"
+CONF_ENTITY_ID = "entity_id"
```
Why not use scene_id?

rokam commented on homeassistant/components/homeassistant/scene.py:
```diff
@@ -123,8 +124,18 @@ def _ensure_no_intersection(value: dict[str, Any]) -> dict[str, Any]:
+    _ensure_no_intersection,
+    vol.Schema(
+        {
+            vol.Required(CONF_ENTITY_ID): cv.entity_id,
```
Why not use scene_id?

emontnemery commented:
@tetele Please mark the PR ready for review once you've addressed the review comments by @rokam 👍

tetele committed:
Split scene validation in 2

First, check if entity_id is a scene
Second, check if it's a scene created with `scene.create`

tetele commented:
@emontnemery @rokam I have preferred `entity_id` to `scene_id` in order to be able to use selectors, autocompleters etc.
Since there was no precedent (there are no other services that delete entities), I just picked what seemed the most reasonable from a US standpoint.

Thoughts? Should I edit the PR and go for `scene_id` or can I leave it as such?

rokam commented:
> @emontnemery @rokam I have preferred `entity_id` to `scene_id` in order to be able to use selectors, autocompleters etc. Since there was no precedent (there are no other services that delete entities), I just picked what seemed the most reasonable from a UX standpoint.
> 
> Thoughts? Should I edit the PR and go for `scene_id` or can I leave it as such?

I think that is better to maintain the same name of the create service. That way it looks more obvious what is expected in this service as a parameter.
Also you can use the entity selector for scene_id. And it might be confusing to expect the scene instead of an entity of other type.
In other hand, entity_id do represent the right term, as scenes are also entities.

So, I don't thing I'm the correct person to decide it.

MartinHjelmare commented:
I've answered this question here:
https://github.com/home-assistant/core/pull/89090#discussion_r1403503494

MartinHjelmare commented on homeassistant/components/homeassistant/scene.py:
```diff
@@ -271,6 +282,28 @@ async def create_service(call: ServiceCall) -> None:
+
+        await platform.async_remove_entity(entity_id)
+
+    hass.services.async_register(
```
This should probably be an entity service as it targets an entity. For those kinds of services we have a helper for registration in the entity platform.

```py
platform = entity_platform.async_get_current_platform()
platform.async_register_entity_service(
    ...
)
```

MartinHjelmare commented on homeassistant/components/scene/services.yaml:
```diff
@@ -54,3 +54,12 @@ create:
         entity:
           multiple: true
+
+delete:
```
I'm not sure why we have the service description for the platform specific services in the scene integration and not under the integration that registered the service. It may be some legacy thing that we haven't updated. Can you try moving this to under the `homeassistant` integration inside a `services.yaml` file?

MartinHjelmare commented on homeassistant/components/scene/strings.json:
```diff
@@ -46,6 +46,16 @@
         }
       }
+    },
+    "delete": {
```
Same here, if it works to move the service yaml descriptions.

MartinHjelmare commented on homeassistant/components/homeassistant/scene.py:
```diff
@@ -271,6 +282,28 @@ async def create_service(call: ServiceCall) -> None:
+            _LOGGER.error("Entity %s is not a scene", entity_id)
+            return
+        if not scene.from_service:
+            _LOGGER.error(
```
Raise `ServiceValidationError` for bad user input and `HomeAssistantError` for other problems. Pass an argument with the message to the exception. The message can have translations and will show in the frontend.

https://developers.home-assistant.io/docs/core/platform/raising_exceptions/

tetele committed:
Address feedback

- Move service to `homeassistant` domain
- Register with `platform.async_register_entity_service`
- Raise validation errors instead of just logging messages

MartinHjelmare commented on tests/components/homeassistant/test_scene.py:
```diff
@@ -164,6 +165,51 @@ async def test_create_service(
+
+    with pytest.raises(ServiceValidationError):
+        await hass.services.async_call(
+            "homeassistant",
```
I remembered why we don't use the platform entity service helper for the `scene.create` service. We want to register the service under the `scene` domain and not under the `homeassistant` domain. We should probably stay consistent with that also for the scene delete service.

MartinHjelmare commented on homeassistant/components/homeassistant/scene.py:
```diff
@@ -271,6 +284,10 @@ async def create_service(call: ServiceCall) -> None:
         SCENE_DOMAIN, SERVICE_CREATE, create_service, CREATE_SCENE_SCHEMA
     )
 
+    platform.async_register_entity_service(
```
Let's move back to using `hass.services.async_register` but create the validation schema using the `cv.make_entity_service_schema` helper. We can do that inline I think.

MartinHjelmare replied:
We can use `async_extract_entity_ids` in the service helper to get the `entity_id`s from the device_ids and area_ids:
https://github.com/home-assistant/core/blob/9ed745638df5e2ae44d5d405a897690518fb6950/homeassistant/helpers/service.py#L433-L440

tetele replied:
But do we want that? Who assigns dynamically created scenes to areas? And do devices even exist in this context?

The service just deletes a small subset of scenes and takes an `entity_id` as an input. What good would this do?

MartinHjelmare replied:
Yeah, I guess it's an edge case. Ok, let's not use `cv.make_entity_service_schema` for now. If we see a need later we can always add it later.

tetele replied:
Too late, I used it and pushed the commit. I guess it won't hurt, although it's certainly overkill.

tetele replied:
I just want to get this (and [this](https://github.com/home-assistant/core/pull/91165)) merged to fix a problem I've been having for almost a year, so if you want the helper, sure, let's engineer the crap of out it :)

MartinHjelmare commented on homeassistant/components/homeassistant/services.yaml:
```diff
@@ -62,3 +62,12 @@ reload_config_entry:
 
 reload_all:
+
+delete_scene:
```
Move this back to under the `scene` integration. Sorry about my confusion.

MartinHjelmare commented on homeassistant/components/homeassistant/strings.json:
```diff
@@ -135,11 +135,24 @@
       "name": "Reload all",
       "description": "Reload all YAML configuration that can be reloaded without restarting Home Assistant."
+    },
+    "delete_scene": {
```
And this.

tetele committed:
Revert moving the service to the `homeassistant` domain

tetele committed:
Remove unneeded validation

tetele committed:
Use helpers and fix tests

MartinHjelmare commented on homeassistant/components/homeassistant/scene.py:
```diff
@@ -271,6 +276,45 @@ async def create_service(call: ServiceCall) -> None:
+        SCENE_DOMAIN,
+        SERVICE_DELETE,
+        delete_service,
+        cv.make_entity_service_schema(
+            {
+                vol.Required(CONF_ENTITY_ID): cv.entity_id,
+            }
+        ),
```
```suggestion
        cv.make_entity_service_schema({}),
```

MartinHjelmare commented on homeassistant/components/scene/services.yaml:
```diff
@@ -54,3 +54,12 @@ create:
           multiple: true
+
+delete:
+  fields:
+    entity_id:
+      required: true
+      example: scene.all_lights
+      selector:
+        entity:
+          domain: scene
```
```suggestion
  target:
    entity:
      - integration: homeassistant
        domain: scene
```

MartinHjelmare commented on tests/components/homeassistant/test_scene.py:
```diff
@@ -164,6 +165,63 @@ async def test_create_service(
+        await hass.async_block_till_done()
+    assert hass.states.get("scene.hallo_2") is not None
+
+    await hass.services.async_call(
```
```suggestion
    assert hass.states.get("scene.hallo") is not None

    await hass.services.async_call(
```

tetele committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

tetele committed:
Fix linting

MartinHjelmare commented:
Are we ready to merge? Please click ready for review when ready.

tetele commented:
I was waiting to see that all checks were passing. Ready for review, cap'n, sir! :D

MartinHjelmare reviewed (approved):
Thanks!
